File: day01/julia/main.jl
function readinput(filename::String)::Tuple{Array{Int64}, Array{Int64}}
    open(filename) do f
        lines = readlines(f)
        left = Array{Int64}(undef, length(lines)) 
        right = Array{Int64}(undef, length(lines)) 
        for (i, l) in enumerate(lines)
            (a, b) = split(l)
            left[i] = parse(Int64, a)
            right[i] = parse(Int64, b)
        end
        
        return (left, right)
    end
end


function part1(left::Array{Int64}, right::Array{Int64})::Int64
    res = 0
    for (a, b) in zip(sort(left), sort(right))
        res += abs(a - b)
    end
    return res
end

function part2(left::Array{Int64}, right::Array{Int64})::Int64
    res = 0
    for a in left
        for b in right
            if a == b
                res += a
            end
        end
    end
    return res
end

function main()
    left, right = readinput("../input.txt")
    println(part1(left, right))
    println(part2(left, right))
end


main()

File: day01/haskell/main.hs
import Data.List (sort)

minAt :: (Ord a) => [a] -> Maybe Int
minAt xs = snd <$> minAt' Nothing (zip xs [0 ..])
  where
    minAt' :: (Ord a) => Maybe (a, Int) -> [(a, Int)] -> Maybe (a, Int)
    minAt' a [] = a
    minAt' (Nothing) ((x, _) : xs) = minAt' (Just (x, 0)) xs
    minAt' (Just (currentMin, minInd)) ((x, ind) : xs) =
        let
            newMin = if x < currentMin then (Just (x, ind)) else (Just (currentMin, minInd))
         in
            minAt' newMin xs

roll :: Int -> [a] -> [a]
roll n xs = drop n xs ++ take n xs

parse :: String -> [(Int, Int)]
parse s =
    let
        numbers = map (map read . words) $ lines s :: [[Int]]
     in
        map (\[a, b] -> (a, b)) numbers

part1 :: [(Int, Int)] -> Int
part1 numbers =
    sum $
        map abs $
            zipWith
                (-)
                (sort $ map fst numbers)
                (sort $ map snd numbers)

count :: (Eq a) => a -> [a] -> Int
count a xs = length $ filter (== a) xs

part2 :: [(Int, Int)] -> Int
part2 numbers =
    let
        left = map fst numbers
        right = map snd numbers
     in
        sum $ map (\n -> (count n right) * n) left

main :: IO ()
main = do
    content <- readFile "../input.txt"
    print $ part1 $ parse content
    print $ part2 $ parse content
